collector/btrfs_linux.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2019 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !nobtrfs
+
+package collector
+
+import (
+	"fmt"
+
+	"github.com/go-kit/kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/procfs/btrfs"
+)
+
+// A btrfsCollector is a Collector which gathers metrics from Btrfs filesystems.
+type btrfsCollector struct {
+	fs     btrfs.FS
+	logger log.Logger
+}
+
+func init() {
+	registerCollector("btrfs", defaultEnabled, NewBtrfsCollector)
+}
+
+// NewBtrfsCollector returns a new Collector exposing Btrfs statistics.
+func NewBtrfsCollector(logger log.Logger) (Collector, error) {
+	fs, err := btrfs.NewFS(*sysPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open sysfs: %v", err)
+	}
+
+	return &btrfsCollector{
+		fs:     fs,
+		logger: logger,
+	}, nil
+}
+
+// Update retrieves and exports Btrfs statistics.
+// It implements Collector.
+func (c *btrfsCollector) Update(ch chan<- prometheus.Metric) error {
+	stats, err := c.fs.Stats()
+	if err != nil {
+		return fmt.Errorf("failed to retrieve Btrfs stats: %v", err)
+	}
+
+	for _, s := range stats {
+		c.updateBtrfsStats(ch, s)
+	}
+
+	return nil
+}
+
+// btrfsMetric represents a single Btrfs metric that is converted into a Prometheus Metric.
+type btrfsMetric struct {
+	name            string
+	desc            string
+	value           float64
+	extraLabel      []string
+	extraLabelValue []string
+}
+
+// updateBtrfsStats collects statistics for one bcache ID.
+func (c *btrfsCollector) updateBtrfsStats(ch chan<- prometheus.Metric, s *btrfs.Stats) {
+	const subsystem = "btrfs"
+
+	// Basic information about the filesystem
+	devLabels := []string{"label", "uuid"}
```

Comment:
These labels are extraneous. Having them on every metric doesn't follow Prometheus best practices. Please include a `btrfs_device_Info` metric that has the `device`, `label` and `uuid` on it.

Reply:
The UUID is not extraneous, as it is the identifier of the filesystem (Btrfs is not limited to a single device).
I see three options:

- `btrfs_info` with both the `uuid` and `label`. The `device`/`uuid` link is already present in `btrfs_device_size_bytes`.
- `btrfs_device_info` containing `uuid`, `label` and `device` (of which there could be several).
- `btrfs_info` with the `uuid` and `label`, and `btrfs_device_info` with the `uuid` and `device`.

I have added a fixup implementing the first one, but I don't really have a preference either way.

Reply:
Ahh, good point about UUID, I had forgotten about how btrfs links devices together.